hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferService.java

```diff
@@ -303,4 +328,130 @@ private void clearRingbuffersHavingLesserBackupCountThan(int partitionId, int th
     public boolean isKnownServiceNamespace(ServiceNamespace namespace) {
         return namespace instanceof ObjectNamespace;
     }
+
+    @Override
+    public Runnable prepareMergeRunnable() {
+        Map<Integer, Map<ObjectNamespace, RingbufferContainer>> partitionContainerMap = createHashMap(containers.size());
+        IPartitionService partitionService = nodeEngine.getPartitionService();
+        Address thisAddress = nodeEngine.getClusterService().getThisAddress();
+
+        for (Entry<Integer, Map<ObjectNamespace, RingbufferContainer>> entry : containers.entrySet()) {
+            int partitionId = entry.getKey();
+            if (!thisAddress.equals(partitionService.getPartitionOwner(partitionId))) {
+                continue;
+            }
+
+            Map<ObjectNamespace, RingbufferContainer> containerMap = partitionContainerMap.get(partitionId);
+            if (containerMap == null) {
+                containerMap = new HashMap<ObjectNamespace, RingbufferContainer>();
+                partitionContainerMap.put(partitionId, containerMap);
+            }
+            for (Entry<ObjectNamespace, RingbufferContainer> containerEntry : entry.getValue().entrySet()) {
+                RingbufferContainer container = containerEntry.getValue();
+                if (!(getMergePolicy(container) instanceof DiscardMergePolicy)) {
+                    containerMap.put(containerEntry.getKey(), container);
+                }
+                container.cleanup();
+            }
+        }
+
+        // clear all items either owned or backup
+        reset();
+
+        return new Merger(partitionContainerMap);
+    }
+
+    private SplitBrainMergePolicy getMergePolicy(RingbufferContainer container) {
+        String mergePolicyName = container.getConfig().getMergePolicyConfig().getPolicy();
+        return mergePolicyProvider.getMergePolicy(mergePolicyName);
+    }
+
+    private class Merger implements Runnable {
+
+        private static final int TIMEOUT_FACTOR = 500;
+
+        private final Map<Integer, Map<ObjectNamespace, RingbufferContainer>> partitionContainerMap;
+
+        Merger(Map<Integer, Map<ObjectNamespace, RingbufferContainer>> partitionContainerMap) {
+            this.partitionContainerMap = partitionContainerMap;
+        }
+
+        @Override
+        public void run() {
+            // we cannot merge into a 3.9 cluster, since not all members may understand the MergeOperation
+            if (nodeEngine.getClusterService().getClusterVersion().isLessThan(Versions.V3_10)) {
```

Comment:
Wondering if we can refactor the `Merger` as well since we seem to be repeating ourselves a bit with other services. We can do this later as well.

Reply:
The logic is similar, but there are subtle differences, depending on the given data structure, the operations being created (e.g. the batching is implemented differently in some data structures) and the log messages etc.